### PR TITLE
Tani multi library v01 - improved beta version

### PR DESCRIPTION
**Pull request by taniwallach**
This is an improved beta version of multi-library support for the library search tools (`SetMaker.pm`) replacing https://github.com/openwebwork/webwork2/pull/874.

 * The library configuration is more complicated than before, so that multiple libraries can be supported. 
 * Each library requires several settings. Several files which read library settings were modified to use the new library config structure.
 * Sample settings are in `conf/localOverrides.conf.dist` 
 * The modified OPL-update script expects the library "code name" as in the config settings, and the second library and on should set the `noDrop` option (to prevent deletion of the database tables shared by all the libraries).
 * The sample config file assumes that the `course/templates` directory will have symlinks to the libraries.
 * Sample test libraries are available at https://github.com/taniwallach/ww-test-libs

Example setup using the test libraries (including OPL update on 3 libraries):
```
# cd into the course directory
cd templates
ln -s /opt/webwork/libraries/tl-01  TestLibrary1
ln -s /opt/webwork/libraries/tl-02  TestLibrary2
ln -s /opt/webwork/libraries/webwork-open-problem-library/OpenProblemLibrary   Library
/opt/webwork/webwork2/bin/OPL-update local OPL
/opt/webwork/webwork2/bin/OPL-update local TSTL1 noDrop
/opt/webwork/webwork2/bin/OPL-update local TSTL2 noDrop
```

Known issue:
 * Do a search on `All Libraries` and it should show a correct count of problems.
 * Then change back to `All Subjects` (the count should update properly).
 * The click on `View Problems`
 * The correct set of problems should be displayed. 
 * *but* the count will be incorrect and the selections will have changes.
 * I currently do not understand how this is happening.

### CODE REVIEW

**Comment by taniwallach** [comment]
This code will apparently need some revision **before** it can be merged into a tree with UTF-8 support, or the merge conflicts will need to be carefully handled.

Once https://github.com/openwebwork/webwork2/pull/934 is merged into develop, I will try (as soon as I can) to prepare create a new PR which gets this code to cleanly merge onto the revised develop branch.

From what I recall, it basically works, and searching against several libraries which were processed by the revised `OPL-update` as explained above works. 

The code is **not complete**.

There are several things which are marked with **FIXME** in the code, and in particular there are places where things are still hardwired to use default values (ex. a sub-path starting with `Library/`) appropriate to the OPL, and which may need to be adjusted to work with different libraries.

In particular more work is needed on:
  * `get_library_sets()` in `lib/WeBWorK/ContentGenerator/Instructor/SetMaker.pm`
  * `getDBextras()`  in `lib/WeBWorK/ContentGenerator/Instructor/SetMaker.pm`
  * `deleteListing()` in `lib/WeBWorK/Utils/ListingDB.pm`
  * `getProblemDirectories()` in `lib/WebworkWebservice/LibraryActions.pm`
but there are almost certainly other parts of the code which require additional work.

Testing of this code by additional people would help.

File: bin/OPL-update [hunk 1]
```diff
@@ -21,6 +21,32 @@ use File::Basename;
 use Cwd;
 use DBI;
 
+# Command line arguments
+# for some reason the first command line argument is disappearing before it gets here
```
Comment (eltenedor) [inline bin/OPL-update:25]: this is caused as `use OPLUtils` seems to cause a shift in the argument, see this excerpt from `OPLUtils.pm` and my comment there

image

File: bin/OPLUtils.pm [hunk 2]
```diff
@@ -25,6 +27,10 @@ use JSON;
 our @EXPORT    = ();
 our @EXPORT_OK = qw(build_library_directory_tree build_library_subject_tree build_library_textbook_tree);
 
+# what library are we handling, now is first argument of the function
+my $myLib = shift;
```
Comment (eltenedor) [inline bin/OPLUtils.pm:31]: I suspect, that this line is not needed.

**Comment by dlglin** [comment]
Closed due to inactivity.  Can be reopened if activity resumes.

**Pull request closed without merging**